### PR TITLE
JIRA2547: include overview of authentication and authorization

### PR DESCRIPTION
**Pull request by chinmayi-chandrasekar**
For 4.9

https://issues.redhat.com/browse/OSDOCS-2547

[Docs preview](https://deploy-preview-39544--osdocs.netlify.app/openshift-enterprise/latest/authentication/index.html)

Requires ack from @xingxingxia @bergerhoffer @s-urbaniak @stlaz

### CODE REVIEW

File: authentication/index.adoc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+* An X.509 client certificate - You require an HTTPS connection to the API server to use the X.509 client certificates for authentication.
+
+[NOTE]
+====
+If you do not present a valid access token or certificate, your request is unauthenticated and you receive an HTTP 401 error.
+====
+An administrator can configure authentication through the following tasks:
+
+* Configure an identity provider: You can define any xref:../authentication/understanding-identity-provider.adoc#supported-identity-providers[supported identity provider in {product-title}] and add it to your cluster. To specify an identity provider, you must create a custom resource (CR) that describes the identity provider.
```
Comment (bergerhoffer) [inline authentication/index.adoc:31]: I don't think that the sentence starting with "To specify..." is important enough to include here, because that's getting into specifics specified elsewhere

File: authentication/index.adoc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+* xref:../authentication/configuring-internal-oauth.adoc#configuring-internal-oauth[Configure the internal OAuth server]: The {product-title} master includes a built-in OAuth server that determines the user’s identity from the configured identity provider and creates access tokens. You can configure the token duration and inactivity timeout, and customize the internal OAuth server URL.
+* xref:../authentication/configuring-oauth-clients.adoc#oauth-register-additional-client_configuring-oauth-clients[Register an OAuth client]: {product-title} includes many xref:../authentication/configuring-oauth-clients.adoc#oauth-default-clients_configuring-oauth-clients[default OAuth clients]. You can register and configure additional OAuth clients.
```
Comment (bergerhoffer) [inline authentication/index.adoc:33]: I'd suggest changing "many" to "several", since there are really only 2 OAuth clients, which I don't think is necessarily many.

File: authentication/index.adoc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+Users must specify either a default or custom OAuth client that receives and uses the token in their requests for the OAuth token.
+====
+
+* Manage cloud provider credentials using the xref:../authentication/managing_cloud_provider_credentials/about-cloud-credential-operator.adoc#about-cloud-credential-operator[Cloud Credentials Operator].
+
+[id="authorization-overview"]
+== Overview of authorization in {product-title}
+Authorization involves determining whether a user has the required permissions to perform an action.
+
+Based on how users interact with an {product-title} cluster, users can be of xref:../authentication/understanding-authentication.adoc#rbac-users_understanding-authentication[three types] - regular users, system users, and service accounts. By default, each user type has a set of access permissions and privileges. These permissions and privileges are represented through the xref:../authentication/using-rbac.adoc#authorization-overview_using-rbac[role-based access control (RBAC) objects], such as rules, roles, and bindings.
```
Comment (bergerhoffer) [inline authentication/index.adoc:46]: Suggest a colon (:) instead of a dash (-) before you list the 3 types of users.

Also suggest rewording "users can be of three types", which sounds a little off to me. Maybe you could rearrange to starting with "There are three types of users... " ?

File: authentication/index.adoc [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+Administrators can manage authorization using the RBAC objects. You can also control access to an {product-title} cluster through xref:../authentication/using-rbac.adoc#rbac-projects-namespaces_using-rbac[projects and namespaces].
+
+An administrator can configure authorization for {product-title} through the following tasks:
+
+* xref:../authentication/using-rbac.adoc#creating-cluster-admin_using-rbac[Create a cluster-admin user]: By default, your cluster has only one cluster administrator called `kubeadmin`. You can create another cluster administrator after defining an identity provider.
```
Comment (bergerhoffer) [inline authentication/index.adoc:52]: When you list the tasks in this list (and also for the task list in the authentication section above), I'd consider starting them all with gerunds (Creating, etc.)

Comment (bergerhoffer) [inline authentication/index.adoc:52]: Also, I'd just check to make sure that it's true that you have to define an IDP before you can create another cluster admin.

File: authentication/index.adoc [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
++
+[NOTE]
+====
+After creating the cluster admin user, xref:../authentication/remove-kubeadmin.adoc#removing-kubeadmin_removing-kubeadmin[delete the existing kubeadmin user] to improve cluster security.
+====
+* xref:../authentication/understanding-and-creating-service-accounts.adoc#service-accounts-managing_understanding-service-accounts[Create service accounts]: xref:../authentication/understanding-and-creating-service-accounts.adoc#service-accounts-overview_understanding-service-accounts[Service accounts] provide a flexible way to control API access without sharing a regular user’s credentials. A user can xref:../authentication/using-service-accounts-in-applications.adoc#service-accounts-using-credentials-externally_using-service-accounts[use a service account in applications] and also as xref:../authentication/using-service-accounts-as-oauth-client.adoc#using-service-accounts-as-oauth-client[an OAuth client]   .
```
Comment (bergerhoffer) [inline authentication/index.adoc:58]: There are back to back links here with "Creating service accounts" and then "Service accounts". I'd suggest to remove the link on "Service accounts" since that content is on the same assembly that you're already linking to with the first link.

Comment (bergerhoffer) [inline authentication/index.adoc:58]: There are extra spaces before the last sentence's period, which is rendering in the build.

File: authentication/index.adoc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+* xref:../authentication/tokens-scoping.adoc#tokens-scoping[Scope tokens]: A scoped token is a token that identifies as a specific user who can perform only specific operations. You can create scoped tokens to delegate some of your permissions to another user or a service account.
+* xref:../authentication/managing-security-context-constraints.adoc#managing-pod-security-policies[Control permissions for pods using security context constraints (SCCs)]: You can use SCCs to define a set of conditions that determine the actions a pod can perform and the resources it can access.
+* Impersonate a system admin user: You can grant cluster administrator permissions to a user by xref:../authentication/impersonating-system-admin.adoc#impersonating-system-admin[impersonating  system admin user].
```
Comment (bergerhoffer) [inline authentication/index.adoc:61]: You need an "a" or "the" in "impersonating  system admin user"

Also, I'm not sure if impersonating the system:admin user is the same thing as "grant cluster administrator permissions to a user", so I'd check with an SME on that.

File: authentication/index.adoc [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+:context: authentication-overview
+
+toc::[]
+
+
+[id="authentication-overview"]
+== Overview of authentication in {product-title}
+To keep an {product-title} cluster secure, a cluster administrator can configure user authentication and ensure only approved users access the cluster.
+
+To interact with an {product-title} cluster, users must authenticate to it by providing one of the following information in their requests to the {product-title} API:
```
Comment (bergerhoffer) [inline authentication/index.adoc:13]: I'd just double check with an SME to see if these are the only two ways

**Review by bergerhoffer — COMMENTED** [review]
Nice job, just a few bits of feedback!

File: authentication/index.adoc [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+Users must specify either a default or custom OAuth client that receives and uses the token in their requests for the OAuth token.
+====
+
+* Managing cloud provider credentials using the xref:../authentication/managing_cloud_provider_credentials/about-cloud-credential-operator.adoc#about-cloud-credential-operator[Cloud Credentials Operator].
+
+[id="authorization-overview"]
+== About authorization in {product-title}
+Authorization involves determining whether a user has the required permissions to perform an action.
+
+Based on how users interact with an {product-title} cluster, there are xref:../authentication/understanding-authentication.adoc#rbac-users_understanding-authentication[three types of users]: regular users, system users, and service accounts. By default, each user type has a set of access permissions and privileges. These permissions and privileges are represented through the xref:../authentication/using-rbac.adoc#authorization-overview_using-rbac[role-based access control (RBAC) objects], such as rules, roles, and bindings.
```
Comment (stlaz) [inline authentication/index.adoc:46]: The distinction is not really based on how the users interact with the cluster, those are just different kinds of users.

File: authentication/index.adoc [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+Administrators can manage authorization using the RBAC objects. You can also control access to an {product-title} cluster through xref:../authentication/using-rbac.adoc#rbac-projects-namespaces_using-rbac[projects and namespaces].
```
Comment (stlaz) [inline authentication/index.adoc:48]: Administrator's can't manage the full extent of authorization only using the RBAC objects. There are SCCs, too, for example, and there might be more mechanisms in the future.

"Administrators can manage user access to cluster resources using the RBAC objects." would perhaps be a more accurate statement?

File: authentication/index.adoc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+An administrator can configure authorization for {product-title} through the following tasks:
```
Comment (stlaz) [inline authentication/index.adoc:50]: I think `manage` is a better term than `configure` when it comes to authorization.

File: authentication/index.adoc [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
++
+[NOTE]
+====
+After creating the cluster admin user, xref:../authentication/remove-kubeadmin.adoc#removing-kubeadmin_removing-kubeadmin[delete the existing kubeadmin user] to improve cluster security.
+====
+* xref:../authentication/understanding-and-creating-service-accounts.adoc#service-accounts-managing_understanding-service-accounts[Creating service accounts]: Service accounts provide a flexible way to control API access without sharing a regular user’s credentials. A user can xref:../authentication/using-service-accounts-in-applications.adoc#service-accounts-using-credentials-externally_using-service-accounts[use a service account in applications] and also as xref:../authentication/using-service-accounts-as-oauth-client.adoc#using-service-accounts-as-oauth-client[an OAuth client].
+* xref:../authentication/tokens-scoping.adoc#tokens-scoping[Scoping tokens]: A scoped token is a token that identifies as a specific user who can perform only specific operations. You can create scoped tokens to delegate some of your permissions to another user or a service account.
```
Comment (stlaz) [inline authentication/index.adoc:59]: Any access token identifies a specific user so that part is redundant.

How do you scope a service account token?

File: authentication/index.adoc [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+* xref:../authentication/managing-security-context-constraints.adoc#managing-pod-security-policies[Controlling permissions for pods using security context constraints (SCCs)]: You can use SCCs to define a set of conditions that determine the actions a pod can perform and the resources it can access.
+* Impersonating a system admin user: You can grant cluster administrator permissions to a user by xref:../authentication/impersonating-system-admin.adoc#impersonating-system-admin[impersonating a system admin user].
```
Comment (stlaz) [inline authentication/index.adoc:61]: I would perhaps classify impersonation as an authentication topic rather than authorization

File: authentication/index.adoc [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+* xref:../authentication/ldap-syncing.adoc#ldap-syncing[Syncing LDAP groups]: You can manage user groups in one place by syncing the groups stored in an LDAP server with the {product-title} user groups.
+* Viewing xref:../authentication/using-rbac.adoc#viewing-local-roles_using-rbac[local] and xref:../authentication/using-rbac.adoc#viewing-cluster-roles_using-rbac[cluster] roles and bindings.
+* Creating xref:../authentication/using-rbac.adoc#creating-local-role_using-rbac[local] and cluster roles, and assigning them to a user or group: {product-title} includes a set of xref:../authentication/using-rbac.adoc#default-roles_using-rbac[default cluster roles]. You can create additional xref:../authentication/using-rbac.adoc#creating-cluster-role_using-rbac[cluster roles] and xref:../authentication/using-rbac.adoc#adding-roles_using-rbac[add them to a user or group].
```
Comment (stlaz) [inline authentication/index.adoc:65]: isn't this just the `Administrators can manage authorization using the RBAC objects` that's mentioned above?

**Review by stlaz — CHANGES_REQUESTED** [review]
A part of this seems to be copied from the "Understanding authentication" section, I don't think we need this kind of redundant information. Consider merging these two.

File: authentication/index.adoc (hunk 9 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:48]: Comment addressed.

File: authentication/index.adoc (hunk 10 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:50]: Comment addressed.

File: authentication/index.adoc (hunk 8 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:46]: Comment addressed.

File: authentication/index.adoc [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+:context: authentication-overview
+
+toc::[]
+
+
+[id="authentication-overview"]
+== About authentication in {product-title}
+To keep an {product-title} cluster secure, a cluster administrator can configure user authentication and ensure only approved users access the cluster.
+
+To interact with an {product-title} cluster, users must authenticate to it by providing one of the following information in their requests to the {product-title} API:
```
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:13]: Comment addressed.

File: authentication/index.adoc [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+= Overview of authentication and authorization
+include::modules/common-attributes.adoc[]
+:context: authentication-overview
+
+toc::[]
+
+
+[id="authentication-overview"]
+== About authentication in {product-title}
+To keep an {product-title} cluster secure, a cluster administrator can configure user authentication and ensure only approved users access the cluster.
```
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:11]: Comment addressed.

File: authentication/index.adoc (hunk 14 above)
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:13]: Comment addressed.

File: authentication/index.adoc [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+====
+If you do not present a valid access token or certificate, your request is unauthenticated and you receive an HTTP 401 error.
+====
+An administrator can configure authentication through the following tasks:
+
+* Configuring an identity provider: You can define any xref:../authentication/understanding-identity-provider.adoc#supported-identity-providers[supported identity provider in {product-title}] and add it to your cluster.
+* xref:../authentication/configuring-internal-oauth.adoc#configuring-internal-oauth[Configuring the internal OAuth server]: The {product-title} control plane includes a built-in OAuth server that determines the user’s identity from the configured identity provider and creates access tokens. You can configure the token duration and inactivity timeout, and customize the internal OAuth server URL.
+* xref:../authentication/configuring-oauth-clients.adoc#oauth-register-additional-client_configuring-oauth-clients[Registering an OAuth client]: {product-title} includes several xref:../authentication/configuring-oauth-clients.adoc#oauth-default-clients_configuring-oauth-clients[default OAuth clients]. You can register and configure additional OAuth clients.
```
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:33]: Comment addressed.

File: authentication/index.adoc [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
++
+[NOTE]
+====
+Users must specify either a default or custom OAuth client that receives and uses the token in their requests for the OAuth token.
+====
```
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:38]: Comment addressed.

File: authentication/index.adoc (hunk 12 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:61]: Comment addressed.

File: authentication/index.adoc (hunk 13 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:65]: Yes, have removed it.

File: authentication/index.adoc [hunk 18]
```diff
@@ -0,0 +1,65 @@
+[id="authentication-authorization-overview"]
+= Overview of authentication and authorization
+include::modules/common-attributes.adoc[]
+:context: authentication-overview
```
[the remark this answers is not in the document]
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:4]: comment addressed.

File: authentication/index.adoc [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+====
+If you do not present a valid access token or certificate, your request is unauthenticated and you receive an HTTP 401 error.
+====
+An administrator can configure authentication through the following tasks:
+
+* Configuring an identity provider: You can define any xref:../authentication/understanding-identity-provider.adoc#supported-identity-providers[supported identity provider in {product-title}] and add it to your cluster.
+* xref:../authentication/configuring-internal-oauth.adoc#configuring-internal-oauth[Configuring the internal OAuth server]: The {product-title} control plane includes a built-in OAuth server that determines the user’s identity from the configured identity provider and creates access tokens. You can configure the token duration and inactivity timeout, and customize the internal OAuth server URL.
+* Registering an OAuth client: {product-title} includes several xref:../authentication/configuring-oauth-clients.adoc#oauth-default-clients_configuring-oauth-clients[default OAuth clients]. You can xref:../authentication/configuring-oauth-clients.adoc#oauth-register-additional-client_configuring-oauth-clients[register and configure additional OAuth clients].
```
Comment (stlaz) [inline authentication/index.adoc:23]: I believe https://docs.openshift.com/container-platform/4.9/authentication/using-service-accounts-as-oauth-client.html belongs here, too

File: authentication/index.adoc [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
++
+[NOTE]
+====
+Users must specify either a default or custom OAuth client that receives and uses the token in their requests for the OAuth token.
```
Comment (stlaz) [inline authentication/index.adoc:27]: This seems confusing, what do you mean?

File: authentication/index.adoc [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+====
+
+* xref:../authentication/managing-oauth-access-tokens.adoc#managing-oauth-access-tokens[Viewing and managing user-owned OAuth access tokens].
```
Comment (stlaz) [inline authentication/index.adoc:30]: this is not necessarily controlled by the administrator but rather the user themselves. Is there a different section where this might fit?

File: authentication/index.adoc [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+You can manage authorization for {product-title} through the following tasks:
+
+* Viewing xref:../authentication/using-rbac.adoc#viewing-local-roles_using-rbac[local] and xref:../authentication/using-rbac.adoc#viewing-cluster-roles_using-rbac[cluster] roles and bindings.
+* Creating a xref:../authentication/using-rbac.adoc#creating-local-role_using-rbac[local role] and assigning it to a user or group.
+* Creating a cluster role and assigning it to a user or group: {product-title} includes a set of xref:../authentication/using-rbac.adoc#default-roles_using-rbac[default cluster roles]. You can create additional xref:../authentication/using-rbac.adoc#creating-cluster-role_using-rbac[cluster roles] and xref:../authentication/using-rbac.adoc#adding-roles_using-rbac[add them to a user or group].
+* Creating a cluster-admin user: By default, your cluster has only one cluster administrator called `kubeadmin`. You can xref:../authentication/using-rbac.adoc#creating-cluster-admin_using-rbac[create another cluster administrator]. Before creating a cluster administrator, ensure that you have configured an identity provider.
```
Comment (stlaz) [inline authentication/index.adoc:49]: There's also the `system:admin` which is the actual real administrator, `kubeadmin` won't work in many scenarios where certain parts of the cluster get broken

The difference between the two is that `kubeadmin` is a more of an OpenShift-kind of a user - you can log in as this user with a username and password. `system:admin` is a kubernetes user that only uses client-certificate authentication. I'm not sure how to get all of this information in here, though.

File: authentication/index.adoc (hunk 20 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:27]: Comment addressed.

File: authentication/index.adoc (hunk 21 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:30]: Comment addressed.

File: authentication/index.adoc (hunk 22 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:49]: @stlaz https://docs.openshift.com/container-platform/4.9/authentication/understanding-identity-provider.html#identity-provider-overview_understanding-identity-provider also mentions just the kubeadmin user. Should we modify the content everywhere?

File: authentication/index.adoc (hunk 19 above)
Comment (chinmayi-chandrasekar) [reply authentication/index.adoc:23]: @stlaz Here, the context is different - Registering an Oauth client is talking about how an admin can configure more OAuth clients for a user.

**Comment by mburke5678** [comment]
/enterprise-4.9
